### PR TITLE
add Bart more output & loss

### PR DESCRIPTION
**Pull request by Yam0214**
### PR types

New features
### PR changes

Models
### Description

add more output and loss for both bart and mbart models

### CODE REVIEW

File: paddlenlp/transformers/bart/modeling.py [hunk 1]
```diff
@@ -523,16 +589,53 @@ def forward(self,
...
+            encoder_output = self.encoder(
+                input_ids,
+                attention_mask,
+                output_attentions=output_attentions,
+                output_hidden_states=output_hidden_states,
+                return_dict=return_dict,
+            )
+        # If the user passed a tuple for encoder_outputs, we wrap it in a BaseModelOutput when return_dict=True
+        elif return_dict and not isinstance(encoder_output, ModelOutput):
+            if isinstance(encoder_output, paddle.Tensor):
```
Comment (FrostML) [inline paddlenlp/transformers/bart/modeling.py]: 这里是否也可以换成上面 `type()` 的方式，比如 `isinstance(encoder_output, type(decoder_input_ids))`，因为如果后面考虑动转静的话，这里的类型就不再是 `paddle.Tensor`，而是一个 variable

Comment (Yam0214) [reply paddlenlp/transformers/bart/modeling.py]: ok

File: paddlenlp/transformers/model_outputs.py [hunk 2]
```diff
@@ -261,6 +342,7 @@ def _transformer_encoder_fwd(self,
 
 # patches of paddle.nn.Transformer to get all hidden_states and attentions
 paddle.nn.TransformerEncoderLayer.forward = _transformer_encoder_layer_fwd
+paddle.nn.TransformerDecoderLayer.forward = _transformer_decoder_layer_fwd
 paddle.nn.TransformerEncoder.forward = _transformer_encoder_fwd
 paddle.nn.TransformerDecoderLayer.forward = _transformer_decoder_layer_fwd
```
Comment (FrostML) [inline paddlenlp/transformers/model_outputs.py]: 这行可以删除，应该也是之前加上相关功能的时候添加的，保留一个就行

File: tests/transformers/mbart/test_modeling.py [hunk 3]
```diff
@@ -214,6 +219,10 @@ def create_and_check_decoder_model_past_large_inputs(
                             atol=1e-3))
 
 
+@parameterized_class(("return_dict", ), [
+    [False],
+    [True],
+])
 class MBartModelTest(ModelTesterMixin, GenerationTesterMixin,
```
Comment (FrostML) [inline tests/transformers/mbart/test_modeling.py:223]: 可以这样设计，在 `testing_utils.py` 文件里面加一个：

``` python
class PaddleNLPModelTest(unittest.TestCase):
    def tearDown(self):
        gc.collect()
```

然后这里继承的 `unittest.TestCase` 改为继承 `PaddleNLPModelTest`，其实相关继承 `unittest.TestCase` 的单测都可以这样作修改，以此来解决爆显存的问题。

File: paddlenlp/transformers/model_outputs.py (hunk 2 above)
Comment (Yam0214) [reply paddlenlp/transformers/model_outputs.py]: 已删除

File: tests/transformers/mbart/test_modeling.py (hunk 3 above)
Comment (Yam0214) [reply tests/transformers/mbart/test_modeling.py:223]: 接受并修改

File: tests/testing_utils.py [hunk 4]
```diff
@@ -18,10 +18,18 @@
...
+import gc
 
 __all__ = ['get_vocab_list', 'stable_softmax', 'cross_entropy']
 
 
+class PaddleNLPModelTest(unittest.TestCase):
+
+    def tearDown(self):
+        gc.collect()
+        return
```
Comment (FrostML) [inline tests/testing_utils.py]: 不用额外加上 `return` 吧

File: tests/transformers/bart/test_modeling.py [hunk 5]
```diff
@@ -443,6 +443,9 @@ def setUp(self):
         np.random.seed(128)
         paddle.seed(128)
 
+    def tearDown(self):
```
Comment (FrostML) [inline tests/transformers/bart/test_modeling.py]: 这里不能直接替换继承的 `unittest.TestCase` 嘛？

**Review by FrostML — APPROVED** [review]

File: paddlenlp/transformers/model_outputs.py [hunk 6]
```diff
@@ -139,45 +107,158 @@ def _transformer_encoder_layer_fwd(self,
...
-                             memory_mask=memory_mask,
-                             cache=None)
+        if self.enable_recompute:
+            outputs = recompute(mod,
+                                tgt,
+                                memory,
+                                tgt_mask=tgt_mask,
+                                memory_mask=memory_mask,
+                                cache=cache[i] if cache else None,
+                                output_attentions=output_attentions)
```
Comment (ZHUI) [inline paddlenlp/transformers/model_outputs.py:217]: 把 cache 的判断为什么删除了？
